// vite-plugin-ssr/dist/cjs/node/plugin/plugins/commonConfig.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.commonConfig = void 0;
const utils_js_1 = require("../utils.js");
const buildConfig_js_1 = require("./buildConfig.js");
const require_shim_1 = require("@brillout/require-shim");
const picocolors_1 = __importDefault(require("@brillout/picocolors"));
const path_1 = __importDefault(require("path"));
const module_1 = require("module");
// @ts-ignore Shimed by dist-cjs-fixup.js for CJS build.
const importMetaUrl = `file://${__filename}`;
const require_ = (0, module_1.createRequire)(importMetaUrl);
function commonConfig() {
    return [
        {
            name: 'vite-plugin-ssr:commonConfig-1',
            config: () => ({
                appType: 'custom',
                ssr: {
                    // Needed as long as VPS is published as CJS.
                    // TODO: can we remove this once VPS is published as ESM?
                    external: ['vite-plugin-ssr', 'vite-plugin-ssr/server']
                }
            }),
            configResolved(config) {
                (0, require_shim_1.installRequireShim_setUserRootDir)(config.root);
            }
        },
        {
            name: 'vite-plugin-ssr:commonConfig-2',
            enforce: 'post',
            configResolved: {
                order: 'post',
                handler(config) {
                    setDefaultPort(config);
                    workaroundCI(config);
                    (0, buildConfig_js_1.assertRollupInput)(config);
                    assertResolveAlias(config);
                    assertEsm(config.root);
                }
            }
        }
    ];
}
exports.commonConfig = commonConfig;
function setDefaultPort(config) {
    var _a, _b;
    // @ts-ignore
    config.server ?? (config.server = {});
    (_a = config.server).port ?? (_a.port = 3000);
    // @ts-ignore
    config.preview ?? (config.preview = {});
    (_b = config.preview).port ?? (_b.port = 3000);
}
// Workaround GitHub Action failing to access the server
function workaroundCI(config) {
    var _a, _b;
    if (process.env.CI) {
        (_a = config.server).host ?? (_a.host = true);
        (_b = config.preview).host ?? (_b.host = true);
    }
}
// TODO/v1-release: replace assertWarning() with assertUsage()
function assertResolveAlias(config) {
    const aliases = getAliases(config);
    const errPrefix = config.configFile || 'Your Vite configuration';
    const errSuffix1 = 'see https://vite-plugin-ssr.com/path-aliases#vite';
    const errSuffix2 = `which will be deprecated in the next major release, use a string insead and ${errSuffix1}`;
    aliases.forEach((alias) => {
        const { customResolver, find } = alias;
        (0, utils_js_1.assertWarning)(customResolver === undefined, `${errPrefix} defines resolve.alias with customResolver() ${errSuffix2}`, { onlyOnce: true });
        if (typeof find !== 'string') {
            (0, utils_js_1.assert)(find instanceof RegExp);
            // Skip aliases set by Vite:
            //   /^\/?@vite\/env/
            //   /^\/?@vite\/client/
            if (find.toString().includes('@vite'))
                return;
            // Skip alias /^solid-refresh$/ set by vite-plugin-solid
            if (find.toString().includes('solid-refresh'))
                return;
            (0, utils_js_1.assertWarning)(false, `${errPrefix} defines resolve.alias with a regular expression ${errSuffix2}`, {
                onlyOnce: true
            });
        }
        else {
            // Skip aliases set by @preact/preset-vite
            if (find.startsWith('react'))
                return;
            (0, utils_js_1.assertWarning)((0, utils_js_1.isValidPathAlias)(find), `${errPrefix} defines an alias ${picocolors_1.default.cyan(find)} that cannot be distinguished from npm package imports, ${errSuffix1}`, { onlyOnce: true });
        }
    });
}
function getAliases(config) {
    const { alias } = config.resolve;
    if (!Array.isArray(alias)) {
        return [alias];
    }
    else {
        return alias;
    }
}
function assertEsm(userViteRoot) {
    const packageJsonPath = (0, utils_js_1.findUserPackageJsonPath)(userViteRoot);
    if (!packageJsonPath)
        return;
    const packageJson = require_(packageJsonPath);
    let dir = path_1.default.dirname(packageJsonPath);
    if (dir !== '/') {
        (0, utils_js_1.assert)(!dir.endsWith('/'));
        dir = dir + '/';
    }
    (0, utils_js_1.assert)(dir.endsWith('/'));
    dir = picocolors_1.default.dim(dir);
    (0, utils_js_1.assertWarning)(packageJson.type === 'module', `We recommend setting ${dir}package.json#type to "module" (and therefore writing ESM code instead of CJS code), see https://vite-plugin-ssr.com/CJS`, { onlyOnce: true });
}
